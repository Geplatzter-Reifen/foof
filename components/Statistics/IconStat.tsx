import React from "react";
import { StyleSheet, View } from "react-native";
import { Text, Icon, TextProps, useTheme } from "@ui-kitten/components";
import { EvaStatus } from "@ui-kitten/components/devsupport";

type IconStatPropType = React.PropsWithChildren<TextProps> & {
  icon: string;
  iconHeight?: number;
  iconWidth?: number;
  fontSize?: number;
  centered?: boolean;
  reversed?: boolean;
};

export default function IconStat(props: IconStatPropType) {
  const iconHeight = props.iconHeight ?? 22;
  const iconWidth = props.iconWidth ?? "auto";
  const fontSize = props.fontSize ?? 18;
  const centered = props.centered ?? false;
  const reversed = props.reversed ?? false;

  // Iconfarbe wird durch den EvaStatus (oder "text") bestimmt
  const theme = useTheme();
  function getColorStringByStatus(iconStatus: EvaStatus): string {
    switch (iconStatus) {
      case "basic":
        return theme["color-basic-400"];
      case "success":
        return theme["color-success-500"];
      case "info":
        return theme["color-info-500"];
      case "warning":
        return theme["color-warning-500"];
      case "danger":
        return theme["color-danger-500"];
      case "text":
        return theme["text-basic-color"];
      default:
        return theme["color-primary-500"];
    }
  }
  const iconColorString = getColorStringByStatus(props.status ?? "primary");

  return (
    <View
      style={
        centered
          ? styles.viewCentered
          : reversed
            ? styles.viewReversed
            : styles.view
      }
      testID="container"
    >
      <Icon
        name={props.icon}
        style={{
          height: iconHeight,
          width: iconWidth,
          color: iconColorString,
          textAlign: "center",
          textAlignVertical: "center",
        }}
      />
      <Text style={{ fontSize: fontSize, ...styles.text }}>
        {props.children}
      </Text>
    </View>
  );
}

const styles = StyleSheet.create({
  view: {
    flexDirection: "row",
  },
  viewReversed: {
    flex: 1,
    flexDirection: "row-reverse",
  },
  viewCentered: {
    flex: 1,
    alignItems: "center",
    gap: 4,
  },
  text: {
    marginHorizontal: 8,
    flexShrink: 1,
    flexWrap: "wrap",
  },
});
