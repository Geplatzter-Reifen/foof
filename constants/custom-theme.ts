export const foofTheme = {
  "color-primary-100": "#FFEDD0",
  "color-primary-200": "#FFD5A1",
  "color-primary-300": "#FFB872",
  "color-primary-400": "#FF9C4E",
  "color-primary-500": "#FF6E14",
  "color-primary-600": "#DB500E",
  "color-primary-700": "#B7370A",
  "color-primary-800": "#932306",
  "color-primary-900": "#7A1403",

  "color-success-100": "#E5FEE1",
  "color-success-200": "#C5FDC4",
  "color-success-300": "#A6FAAC",
  "color-success-400": "#8EF59F",
  "color-success-500": "#69EF8D",
  "color-success-600": "#4CCD7B",
  "color-success-700": "#34AC6C",
  "color-success-800": "#218A5C",
  "color-success-900": "#147252",

  "color-info-100": "#DFFAD7",
  "color-info-200": "#BAF5B1",
  "color-info-300": "#87E183",
  "color-info-400": "#5EC363",
  "color-info-500": "#309c40",
  "color-info-600": "#23863B",
  "color-info-700": "#187035",
  "color-info-800": "#0F5A2F",
  "color-info-900": "#094A2B",

  "color-warning-100": "#FFF9D9",
  "color-warning-200": "#FFF1B4",
  "color-warning-300": "#FFE78E",
  "color-warning-400": "#FFDD72",
  "color-warning-500": "#FFCD44",
  "color-warning-600": "#DBA931",
  "color-warning-700": "#B78722",
  "color-warning-800": "#936715",
  "color-warning-900": "#7A510D",

  "color-danger-100": "#FFEBDD",
  "color-danger-200": "#FFD2BB",
  "color-danger-300": "#FFB499",
  "color-danger-400": "#FF9780",
  "color-danger-500": "#FF6756",
  "color-danger-600": "#DB413E",
  "color-danger-700": "#B72B33",
  "color-danger-800": "#931B2C",
  "color-danger-900": "#7A1028",

  "color-basic-100": "#FFFFFF",
  "color-basic-200": "#eaeaea",
  "color-basic-300": "#b8b8b8",
  "color-basic-400": "#7e7e7e",
  "color-basic-500": "#494949",
  "color-basic-600": "#373737",
  "color-basic-700": "#282828",
  "color-basic-800": "#202020",
  "color-basic-900": "#171717",
  "color-basic-1000": "#101010",
  "color-basic-1100": "#000000",
};

export const foofLightTheme = {
  ...foofTheme,

  "color-basic-transparent-100": "rgba(126, 126, 126, 0.08)",
  "color-basic-transparent-200": "rgba(126, 126, 126, 0.16)",
  "color-basic-transparent-300": "rgba(126, 126, 126, 0.24)",
  "color-basic-transparent-400": "rgba(126, 126, 126, 0.32)",
  "color-basic-transparent-500": "rgba(126, 126, 126, 0.4)",
  "color-basic-transparent-600": "rgba(126, 126, 126, 0.48)",

  // Custom background color
  "background-basic-color-1": "$color-basic-100",
  "background-basic-color-2": "$color-basic-200",
  "background-basic-color-3": "$color-basic-300",
  "background-basic-color-4": "$color-basic-400",

  // Custom text colors
  "text-basic-color": "#000000",
  "text-hint-color": "$color-basic-600",
};

export const foofDarkTheme = {
  ...foofTheme,

  "color-basic-transparent-100": "rgba(47, 40, 40, 0.08)",
  "color-basic-transparent-200": "rgba(47, 40, 40, 0.16)",
  "color-basic-transparent-300": "rgba(47, 40, 40, 0.24)",
  "color-basic-transparent-400": "rgba(47, 40, 40, 0.32)",
  "color-basic-transparent-500": "rgba(47, 40, 40, 0.4)",
  "color-basic-transparent-600": "rgba(47, 40, 40, 0.48)",

  // Custom background color
  "background-basic-color-1": "$color-basic-600",
  "background-basic-color-2": "$color-basic-700",
  "background-basic-color-3": "$color-basic-800",
  "background-basic-color-4": "$color-basic-900",

  // Custom text colors
  "text-basic-color": "#FFFFFF",
  "text-hint-color": "$color-basic-500",
};
