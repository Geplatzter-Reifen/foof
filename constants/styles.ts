import { StyleSheet } from "react-native";

const customStyles = StyleSheet.create({
  basicCard: {
    borderRadius: 5,
  },
  basicShadow: {
    // iOS
    shadowColor: "#000",
    shadowOffset: { width: 0, height: 1 },
    shadowOpacity: 0.2,
    shadowRadius: 1.5,

    // Android
    elevation: 2,
  },
});

export default customStyles;
