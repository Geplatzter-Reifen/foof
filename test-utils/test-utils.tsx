// https://testing-library.com/docs/react-testing-library/setup/#add-custom-queries
import React from "react";
import { render, RenderOptions } from "@testing-library/react-native";
import { FontAwesomeIconsPack } from "@/components/Font/fontAwesome";
import { mapping } from "@eva-design/eva";
import {
  IconRegistry,
  ApplicationProvider,
  ModalService,
} from "@ui-kitten/components";
import * as eva from "@eva-design/eva";
import { foofDarkTheme, foofLightTheme } from "@/constants/custom-theme";
import { library } from "@fortawesome/fontawesome-svg-core";
import { far } from "@fortawesome/free-regular-svg-icons";
import { fas } from "@fortawesome/free-solid-svg-icons";
import { fab } from "@fortawesome/free-brands-svg-icons";

// Iconpacks laden
library.add(far, fas, fab);

// ModalService wird verwendet, um die Modal-Positionierung zu steuern
ModalService.setShouldUseTopInsets = true;

type Props = {
  children: React.ReactNode;
  use_dark_theme?: boolean;
};
// Der Wrapper wird verwendet, um die Provider zu umschließen, die für die Komponenten notwendig sind
const AllTheProviders = ({ children, use_dark_theme = true }: Props) => {
  const theme = use_dark_theme
    ? { ...eva.dark, ...foofDarkTheme }
    : { ...eva.light, ...foofLightTheme };
  return (
    <>
      <IconRegistry icons={FontAwesomeIconsPack} />
      <ApplicationProvider {...eva} customMapping={mapping} theme={theme}>
        {children}
      </ApplicationProvider>
    </>
  );
};

// Erstellt einen Custom-Render, der die Provider um die Komponenten wickelt
const customRender = (
  component: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => {
  return render(component, { wrapper: AllTheProviders, ...options });
};

// Exportiert alle Funktionen der Testing Library, weswegen man diese dann auch von dieser Datei importieren kann
export * from "@testing-library/react-native";
export { customRender as render };
