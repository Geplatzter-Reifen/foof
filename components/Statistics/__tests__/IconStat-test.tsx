import { render } from "@/test-utils/test-utils";
import IconStat from "@/components/Statistics/IconStat";
import { foofLightTheme, foofTheme } from "@/constants/custom-theme";

describe("TourStats", () => {
  it("should render the correct icon", async () => {
    const view = render(<IconStat icon="home">1001km</IconStat>);
    expect(view).toMatchSnapshot();
  });
  it("should render the icon and font in correct size", () => {
    const view = render(
      <IconStat icon="home" iconSize={30} fontSize={21}>
        1001km
      </IconStat>,
    );
    expect(view).toMatchSnapshot();
  });
  describe("should render the correct icon color depending on the given status", () => {
    test("primary", () => {
      const { getByTestId } = render(<IconStat icon="home" status="primary" />);
      const view = getByTestId("container");
      const icon = view.props.children[0];
      expect(icon.props.style.color).toBe(foofTheme["color-primary-500"]);
    });
    test("basic", () => {
      const { getByTestId } = render(<IconStat icon="home" status="basic" />);
      const view = getByTestId("container");
      const icon = view.props.children[0];
      expect(icon.props.style.color).toBe(foofTheme["color-basic-500"]);
    });
    test("success", () => {
      const { getByTestId } = render(<IconStat icon="home" status="success" />);
      const view = getByTestId("container");
      const icon = view.props.children[0];
      expect(icon.props.style.color).toBe(foofTheme["color-success-500"]);
    });
    test("info", () => {
      const { getByTestId } = render(<IconStat icon="home" status="info" />);
      const view = getByTestId("container");
      const icon = view.props.children[0];
      expect(icon.props.style.color).toBe(foofTheme["color-info-500"]);
    });
    test("warning", () => {
      const { getByTestId } = render(<IconStat icon="home" status="warning" />);
      const view = getByTestId("container");
      const icon = view.props.children[0];
      expect(icon.props.style.color).toBe(foofTheme["color-warning-500"]);
    });
    test("danger", () => {
      const { getByTestId } = render(<IconStat icon="home" status="danger" />);
      const view = getByTestId("container");
      const icon = view.props.children[0];
      expect(icon.props.style.color).toBe(foofTheme["color-danger-500"]);
    });
    test("text", () => {
      const { getByTestId } = render(<IconStat icon="home" status="text" />);
      const view = getByTestId("container");
      const icon = view.props.children[0];
      expect(icon.props.style.color).toBe(foofLightTheme["text-basic-color"]); // wÃ¤re im Darkmode das andere Theme
    });
    test("incorrect syntax should render primary color", () => {
      const { getByTestId } = render(<IconStat icon="home" status="xyz" />);
      const view = getByTestId("container");
      const icon = view.props.children[0];
      expect(icon.props.style.color).toBe(foofTheme["color-primary-500"]);
    });
    test("no given status should render primary color", () => {
      const { getByTestId } = render(<IconStat icon="home" />);
      const view = getByTestId("container");
      const icon = view.props.children[0];
      expect(icon.props.style.color).toBe(foofTheme["color-primary-500"]);
    });
  });
});
