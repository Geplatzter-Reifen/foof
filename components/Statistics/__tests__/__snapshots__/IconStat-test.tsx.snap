// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TourStats should render the correct icon 1`] = `
<View
  style={
    {
      "flexDirection": "row",
    }
  }
  testID="container"
>
  <View
    collapsable={false}
    style={
      {
        "transform": [
          {
            "scale": 1,
          },
        ],
      }
    }
  >
    <Text />
  </View>
  <Text
    appearance="default"
    category="p1"
    status="basic"
    style={
      [
        {
          "color": "#000000",
          "fontFamily": "System",
          "fontSize": 15,
          "fontWeight": "400",
        },
        {
          "flexShrink": 1,
          "flexWrap": "wrap",
          "fontSize": 18,
          "marginHorizontal": 8,
        },
      ]
    }
  >
    1001km
  </Text>
</View>
`;

exports[`TourStats should render the icon and font in correct size 1`] = `
<View
  style={
    {
      "flexDirection": "row",
    }
  }
  testID="container"
>
  <View
    collapsable={false}
    style={
      {
        "transform": [
          {
            "scale": 1,
          },
        ],
      }
    }
  >
    <Text
      allowFontScaling={false}
      selectable={false}
      style={
        [
          {
            "color": undefined,
            "fontSize": 30,
          },
          {
            "color": "#FF6E14",
            "textAlign": "center",
            "textAlignVertical": "center",
            "width": 35,
          },
          {
            "fontFamily": "FontAwesome6Free-Regular",
            "fontStyle": "normal",
            "fontWeight": "normal",
          },
          {
            "fontWeight": "400",
          },
        ]
      }
    >
      ï€•
    </Text>
  </View>
  <Text
    appearance="default"
    category="p1"
    status="basic"
    style={
      [
        {
          "color": "#000000",
          "fontFamily": "System",
          "fontSize": 15,
          "fontWeight": "400",
        },
        {
          "flexShrink": 1,
          "flexWrap": "wrap",
          "fontSize": 21,
          "marginHorizontal": 8,
        },
      ]
    }
  >
    1001km
  </Text>
</View>
`;
